### PR TITLE
Allow for gracefully completely disabling build scans

### PR DESCRIPTION
**Pull request by ldaley**
… when Gradle knows its not compatible with the scan plugin version.

This is driven by scan functionality currently not working in the presence of source mappings. Previously, the scan plugin was reaching through internal and public incubating API, making it public API in effect, in order to work out if there were VCS mappings. Instead, we now use our existing insulation layer to convey this.

Moreover, for future situations like this, it is now possible for Gradle to tell the scan plugin that it should not try and do its thing. Had this already existed, we could have used it here. Said another way, we can use this to avoid hard failures when we know something about a new version of Gradle makes it incompatible with existing scan plugin versions, which we are likely to know before we release the version of Gradle in question.

CI: https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=ldaley%2Funsupported-scan-plugin-coordination

### CODE REVIEW

**Comment by breskeby** [comment]
looks good to me. 
just to double check, So the general approach is to keep checks in the plugin that allow the plugin to fail the build gracefully (e.g. explicit composite builds not supported atm) and keep everything that needs to captured in advance here?

File: subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy [hunk 1]
```diff
@@ -57,43 +59,92 @@ class BuildScanConfigManagerTest extends Specification {
...
+        attributes.isRootProjectHasVcsMappings() >> true
+
+        when:
+        config("1.9")
+
+        then:
+        thrown UnsupportedBuildScanPluginVersionException
+    }
+
+    def "does not throw if has vcs mappings and plugin version 1.10 or newer"() {
```
Comment (breskeby) [inline subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy]: this test doesn't test version 1.10 as described in the title

File: subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy [hunk 2]
```diff
@@ -57,43 +59,92 @@ class BuildScanConfigManagerTest extends Specification {
...
+        config("1.11")
+
+        then:
+        notThrown UnsupportedBuildScanPluginVersionException
+
+        when:
+        config("1.12")
+
+        then:
+        notThrown UnsupportedBuildScanPluginVersionException
```
Comment (breskeby) [inline subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy]: I'd use use spock where clause here

**Comment by ldaley** [comment]
@breskeby yeah, there are two mechanisms in this:

1. Tell the scan plugin to not bother because Gradle knows its not going to work (generally useful for already released versions)
2. Tell the scan plugin about things that we know it uses to decide whether to apply or not (useful for the next version where we update the plugin to look for this)

1) is there as a way to prevent things from blowing up with obscure failures as was the case with the VCS mappings for existing versions.

2) Is there as a general way to convey whatever we need. For 1.11 version of the plugin we'll use this to see if there are VCS mappings and disable ourselves on our side. Later, if this is no longer a problem for us, we can decide to go ahead and work.

File: subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy (hunk 1 above)
Comment (ldaley) [reply subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy]: Done.

File: subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy (hunk 2 above)
Comment (ldaley) [reply subprojects/core/src/test/groovy/org/gradle/internal/scan/config/BuildScanConfigManagerTest.groovy]: Done.

**Comment by ldaley** [comment]
@eljobe @oehme CS has finished reviewing this. We are waiting on BT team sign off to merge.

**Comment by ldaley** [comment]
@oehme @eljobe ping.

**Comment by ldaley** [comment]
We need to move on here. I'm optimistically merging this.

**Pull request merged**